cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/IncentiveSystem.java

```diff
@@ -33,28 +46,128 @@ public UserKeyPair generateUserKeys() {
+
+    /**
+     * implementation of the Issue<->Join protocol
+     */
+
+
+    /**
+     * functionality of the first part of the Issue algorithm of the Cryptimeleon incentive system
+     * @param pp public parameters of the respective incentive system instance
+     * @param pk provider public key of the provider the user interacts with
+     * @param ukp user key pair
+     * @return join request, i.e. object representing the first two messages in the Issue-Join protocol of the Cryptimeleon incentive system
+     */
+    public JoinRequest generateJoinRequest(IncentivePublicParameters pp, ProviderPublicKey pk, UserKeyPair ukp, ZnElement eskUsr, ZnElement dsrnd0, ZnElement dsrnd1, ZnElement z, ZnElement t, ZnElement u) {
+        UserPublicKey upk = ukp.getPk();
+        UserSecretKey usk = ukp.getSk();
+
+        // TODO: generate random values needed for generation of fresh user token using PRF
+        //  (currently, they are passed as method parameters until PRF stuff has been figured out),
+        //  use Paul's stuff (see notes)
+
+        // compute Pedersen commitment for user token
+        RingElementVector exponents = new RingElementVector(usk.getUsk(), eskUsr, dsrnd0, dsrnd1, pp.getBg().getZn().getZeroElement(), z); // need to retrieve exponent from usk object; point count of 0 is reresented by zero in used Z_n
+        GroupElement c0Pre = pk.getH().innerProduct(exponents).op(pp.getH7().pow(t)).pow(u);
+        GroupElement c1Pre = pp.getG1Generator().pow(u);
+
+        // compute NIZKP to prove well-formedness of token
+        FiatShamirProofSystem cwfProofSystem = new FiatShamirProofSystem(new CommitmentWellformednessProtocol(pp, pk));
+        CommitmentWellformednessCommonInput cwfCommon = new CommitmentWellformednessCommonInput(upk.getUpk(), c0Pre, c1Pre);
+        CommitmentWellformednessWitness cwfWitness = new CommitmentWellformednessWitness(usk.getUsk(), eskUsr, dsrnd0, dsrnd1, z, t, u.inv());
+        FiatShamirProof cwfProof = cwfProofSystem.createProof(cwfCommon, cwfWitness);
+
+        // assemble and return join request object (commitment, proof of well-formedness)
+        return new JoinRequest(c0Pre, c1Pre, cwfProof);
     }
 
-    void handleJoinRequestResponse() {
+    /**
+     * Implements the functionality of the Issue algorithm of the Cryptimeleon incentive system, i.e. handles a join request by signing the
+     * included preliminary commitment after adding the provider's share for the tracking key esk.
+     * @param pp public parameters of the respective incentive system instance
+     * @param pkp key pair of the provider
+     * @param upk public key of user (needed to restore upk needed to check validity of the commitment well-formedness proof)
+     * @param jr join request to be handled
+     * @return join response, i.e. object representing the third message in the Issue-Join protocol
+     * @throws IllegalArgumentException indicating that the proof for commitment well-formedness was rejected
```

Comment:
What would be the benefit of a runtime exception here? (that it indicates that the faulty parameter cannot be identified without executing some check method?)
And why is defining custom exceptions fine here?
I recall a discussion we had in January where the point was that IllegalArgumentExceptions are preferred over custom ones. It was about the PedersenException indicating that something was wrong with the base group element vector h in the provider public key.

Reply:
I am referring to the server-side exception handling/control flow. If we require catching an exception, I would prefer having our own exception for this use case (e.g. InvalidRequestException). Or make this a runtime exception and assume this never happens (why should it if users know their requests are verified. In this case, I would not require catching that exception.

IMO you can leave it as it is and instead open an issue since the same applies to the other protocols

Reply:
See https://github.com/cryptimeleon/incentive-system/issues/32